HW3/P4/median_filter.cl

```diff
@@ -12,23 +13,81 @@ median_3x3(__global __read_only float *in_values,
     // Note: It may be easier for you to implement median filtering
     // without using the local buffer, first, then adjust your code to
     // use such a buffer after you have that working.
+    int gID, lID, x, y, lx, ly, gSizeX, gSizeY, 
+        lSizeX, lSizeY, xTemp, yTemp, xUse, yUse,
+        buf_corner_x, buf_corner_y, buf_x, buf_y, row;
+    // the code below is adapted from the lecture code on halos
+    x = get_global_id(0);
+    y = get_global_id(1);
+    lx = get_local_id(0);
+    ly = get_local_id(1);
+    gSizeX = get_global_size(0);
+    gSizeY = get_global_size(1);
+    lSizeX = get_local_size(0);
+    lSizeY = get_local_size(1);
+    
+    
+    gID = gSizeX*y + x;
+    lID = lSizeX*ly + lx;
 
+    buf_corner_x = x - lx - halo;
+    buf_corner_y = y - ly - halo;
 
-    // Load into buffer (with 1-pixel halo).
-    //
-    // It may be helpful to consult HW3 Problem 5, and
-    // https://github.com/harvard-cs205/OpenCL-examples/blob/master/load_halo.cl
-    //
-    // Note that globally out-of-bounds pixels should be replaced
-    // with the nearest valid pixel's value.
+    buf_x = lx + halo;
+    buf_y = ly + halo;
 
+    if ((y < h) && (x < w)){
+        if (lID < buf_w){ // only work with buf_w threads
+            xTemp = buf_corner_x + lID;
+            xUse = xTemp;
+            if (xTemp < 0){ // if pixel out of bounds, add compensation steps to find closest in bound pixel
+                    xUse += 1;
+            }
+            if (xTemp > w - 1){
+                xUse -= 1;
```

Comment:
This would lead to accessing wrong values if the buffer size has more than 1 value out of bounds with the left side of the input matrix.